src/phylum/ci/ci_azure.py

```diff
@@ -0,0 +1,229 @@
+"""Define an implementation for the Azure Pipelines CI platform.
+
+Azure References:
+  * https://learn.microsoft.com/azure/devops/pipelines
+  * https://learn.microsoft.com/azure/devops/pipelines/process/access-tokens
+  * https://learn.microsoft.com/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate
+  * https://learn.microsoft.com/azure/devops/pipelines/build/variables
+  * https://learn.microsoft.com/azure/devops/pipelines/process/set-secret-variables
+  * https://learn.microsoft.com/azure/devops/pipelines/yaml-schema
+  * https://learn.microsoft.com/azure/devops/pipelines/yaml-schema/steps-checkout
+  * https://learn.microsoft.com/rest/api/azure/devops
+  * https://learn.microsoft.com/rest/api/azure/devops/git/pull-request-threads/list
+  * https://learn.microsoft.com/rest/api/azure/devops/git/pull-request-threads/create
+"""
+import base64
+import os
+import subprocess
+import sys
+import urllib.parse
+from argparse import Namespace
+from pathlib import Path
+from typing import Optional
+
+import requests
+from phylum.ci.ci_base import CIBase, git_remote
+from phylum.ci.constants import PHYLUM_HEADER
+from phylum.constants import REQ_TIMEOUT
+
+
+class CIAzure(CIBase):
+    """Provide methods for an Azure Pipelines CI environment."""
+
+    def __init__(self, args: Namespace) -> None:
+        super().__init__(args)
+        self.ci_platform_name = "Azure Pipelines"
+
+    def _check_prerequisites(self) -> None:
+        """Ensure the necessary pre-requisites are met and bail when they aren't.
+
+        These are the current pre-requisites for operating within an Azure Pipelines Environment:
+          * The environment must actually be within Azure Pipelines
+          * The environment must be part of a Pull Request (PR) pipeline
+          * An Azure token providing API access is available
+        """
+        super()._check_prerequisites()
+
+        # References:
+        # https://github.com/watson/ci-info/blob/master/vendors.json
+        # https://learn.microsoft.com/azure/devops/pipelines/build/variables
+        if os.getenv("SYSTEM_TEAMFOUNDATIONCOLLECTIONURI") is None:
+            raise SystemExit(" [!] Must be working within the Azure Pipelines environment")
+        if os.getenv("SYSTEM_PULLREQUEST_PULLREQUESTID") is None:
+            print(" [+] Not in a Pull Request pipeline. Nothing to do. Exiting ...")
+            sys.exit(0)
```

Comment:
Yes, this can happen in a properly configured pipeline. Here is an example of it happening in the test repo used to verify the functionality of this new integration:

<img width="1394" alt="image" src="https://user-images.githubusercontent.com/18729796/193075638-01397a3e-883f-4c60-bdd6-dc132654b89a.png">

That pipeline run was triggered by a push to the `main` branch when the PR for the test branch was completed. This is known as a "CI trigger." There are also "PR triggers" but they have no affect for Azure Repos Git based repositories. For more info, see [this documentation about triggers](https://learn.microsoft.com//azure/devops/pipelines/build/triggers). Following that, to the specific use case supported by this integration (Azure Repos Git hosted repositories) it [can be seen](https://learn.microsoft.com/azure/devops/pipelines/repos/azure-repos-git#pr-triggers) that this PR trigger functionality is implemented using branch policies instead.

All that said, a new issue is forthcoming, to allow for so called "build pipelines" in addition to "PR pipelines." This feature will essentially allow a build (push to any branch...not tied to a PR) to pass/fail based on the Phylum analysis, extending the support past PRs only. This feature was directed via @furi0us333 and comes from a user request on the GitLab integration (GitLab and Azure Pipelines both work this way). Basically, the `sys.exit(0)` call will be removed and a proper return code will be given, after performing the analysis. Output (PR comments) will also be skipped.